packages/ui/VirtualList/VirtualListBase.js

```diff
@@ -48,446 +47,322 @@ const itemSizesShape = PropTypes.shape({
+		const itemContainerRef = useRef();
 
-			super(props);
+		const [prevChildProps, setPrevChildProps] = useState(null);
+		const [prevFirstIndex, setPrevFirstIndex] = useState(0);
+		const [updateFrom, setUpdateFrom] = useState(0);
+		const [updateTo, setUpdateTo] = useState(0);
+		const [firstIndex, setFirstIndex] = useState(0);
+		const [numOfItems, setNumOfItems] = useState(0);
 
-			this.containerRef = React.createRef();
-			this.contentRef = React.createRef();
-			this.itemContainerRef = React.createRef();
-
-			if (props.clientSize) {
-				this.calculateMetrics(props);
-				nextState = this.getStatesAndUpdateBounds(props);
-			}
-
-			this.state = {
-				firstIndex: 0,
-				numOfItems: 0,
-				prevChildProps: null,
-				prevFirstIndex: 0,
-				updateFrom: 0,
-				updateTo: 0,
-				...nextState
-			};
+		if (props.clientSize) {
+			calculateMetrics(props);
+			nextState = getStatesAndUpdateBounds();
+			setFirstIndex(nextState.firstIndex);
+			setNumOfItems(nextState.numOfItems);
 		}
 
-		static getDerivedStateFromProps (props, state) {
+		// Instance variables
+		const variables = useRef({
+			scrollBounds: {
+				clientWidth: 0,
+				clientHeight: 0,
+				scrollWidth: 0,
+				scrollHeight: 0,
+				maxLeft: 0,
+				maxTop: 0
+			},
+
+			moreInfo: {
+				firstVisibleIndex: null,
+				lastVisibleIndex: null
+			},
+
+			primary: null,
+			secondary: null,
+
+			isPrimaryDirectionVertical: true,
+			isItemSized: false,
+
+			shouldUpdateBounds: false,
+
+			dimensionToExtent: 0,
+			threshold: 0,
+			maxFirstIndex: 0,
+			curDataSize: 0,
+			hasDataSizeChanged: false,
+			cc: [],
+			scrollPosition: 0,
+			scrollPositionTarget: 0,
+
+			// For individually sized item
+			itemPositions: [],
+			indexToScrollIntoView: -1,
+
+			deferScrollTo : false
+		});
+
+		useImperativeHandle(reference, () => ({
+			calculateMetrics,
+			containerRef,
+			get dimensionToExtent () {
+				return variables.current.dimensionToExtent;
+			},
+			getGridPosition,
+			getItemBottomPosition,
+			getItemNode,
+			getItemPosition,
+			getMoreInfo,
+			getScrollBounds,
+			gridPositionToItemPosition,
+			isHorizontal,
+			get isPrimaryDirectionVertical () {
+				return variables.current.isPrimaryDirectionVertical;
+			},
+			isVertical,
+			get itemPositions () {
+				return variables.current.itemPositions;
+			},
+			get primary () {
+				return variables.current.primary;
+			},
+			props,
+			get scrollPositionTarget () {
+				return variables.current.scrollPositionTarget;
+			},
+			state: {
+				numOfItems
+			},
+			get scrollPosition () {
+				return variables.current.scrollPosition;
+			},
+			scrollToPosition,
+			setScrollPosition,
+			syncClientSize
+		}));
+
+		// getDerivedStateFromProps
+		useLayoutEffect(() => {
 			const
 				shouldInvalidate = (
-					state.prevFirstIndex === state.firstIndex ||
-					state.prevChildProps !== props.childProps
+					prevFirstIndex === firstIndex ||
+					prevChildProps !== props.childProps
```

Comment:
This is not a bug of this PR, but let's reconsider whether this comparison is actually meaningful. In a normal render from Scrollable, `childProps` is re-generated.
And, it looks that `prevFirstIndex` and `prevChildProps` have a similar pattern; those values are accessed only in this effect. We can put those values into a single state object if we have a better outcome. Please consider this.

Reply:
> let's reconsider whether this comparison is actually meaningful.
I couldn't trim that logic currently. so I just left description on this code. 

>  it looks that prevFirstIndex and prevChildProps have a similar pattern; ...
I don't think these variables have to be state. I moved these variable to `variables.current`.